### PR TITLE
replaced all API facing `int`'s with `int32_t`

### PR DESCRIPTION
**Pull request by MarcusDunn**
Easy part of #4574

### CODE REVIEW

File: llama.cpp [hunk 1]
```diff
@@ -7512,7 +7512,7 @@ void llama_sample_softmax(struct llama_context * ctx, llama_token_data_array * c
     }
 }
 
-void llama_sample_top_k(struct llama_context * ctx, llama_token_data_array * candidates, int k, size_t min_keep) {
+void llama_sample_top_k(struct llama_context * ctx, llama_token_data_array * candidates, int32_t k, size_t min_keep) {
```
Comment (LoganDark) [inline llama.cpp:7515]: size_t

File: llama.cpp [hunk 2]
```diff
@@ -7872,7 +7872,7 @@ void llama_sample_classifier_free_guidance(
     }
 }
 
-llama_token llama_sample_token_mirostat(struct llama_context * ctx, llama_token_data_array * candidates, float tau, float eta, int m, float * mu) {
+llama_token llama_sample_token_mirostat(struct llama_context * ctx, llama_token_data_array * candidates, float tau, float eta, int32_t m, float * mu) {
```
Comment (LoganDark) [inline llama.cpp:7875]: size_t

File: llama.cpp [hunk 3]
```diff
@@ -9383,23 +9383,23 @@ enum llama_vocab_type llama_vocab_type(const struct llama_model * model) {
     return model->vocab.type;
 }
 
-int llama_n_vocab(const struct llama_model * model) {
+int32_t llama_n_vocab(const struct llama_model * model) {
```
Comment (LoganDark) [inline llama.cpp:9386]: size_t

File: llama.cpp [hunk 4]
```diff
@@ -9383,23 +9383,23 @@ enum llama_vocab_type llama_vocab_type(const struct llama_model * model) {
...
     return model->vocab.id_to_token.size();
 }
 
-int llama_n_ctx_train(const struct llama_model * model) {
+int32_t llama_n_ctx_train(const struct llama_model * model) {
```
Comment (LoganDark) [inline llama.cpp:9390]: size_t

File: llama.cpp [hunk 5]
```diff
@@ -9383,23 +9383,23 @@ enum llama_vocab_type llama_vocab_type(const struct llama_model * model) {
...
     return model->hparams.n_ctx_train;
 }
 
-int llama_n_embd(const struct llama_model * model) {
+int32_t llama_n_embd(const struct llama_model * model) {
```
Comment (LoganDark) [inline llama.cpp:9394]: size_t

File: llama.cpp [hunk 6]
```diff
@@ -9383,23 +9383,23 @@ enum llama_vocab_type llama_vocab_type(const struct llama_model * model) {
...
     return model->hparams.n_embd;
 }
 
 float llama_rope_freq_scale_train(const struct llama_model * model) {
     return model->hparams.rope_freq_scale_train;
 }
 
-int llama_model_meta_val_str(const struct llama_model * model, const char * key, char * buf, size_t buf_size) {
+int32_t llama_model_meta_val_str(const struct llama_model * model, const char * key, char * buf, size_t buf_size) {
```
Comment (LoganDark) [inline llama.cpp:9402]: ssize_t

File: llama.cpp [hunk 7]
```diff
@@ -9410,11 +9410,11 @@ int llama_model_meta_val_str(const struct llama_model * model, const char * key,
     return snprintf(buf, buf_size, "%s", it->second.c_str());
 }
 
-int llama_model_meta_count(const struct llama_model * model) {
+int32_t llama_model_meta_count(const struct llama_model * model) {
```
Comment (LoganDark) [inline llama.cpp:9413]: size_t

File: llama.cpp [hunk 8]
```diff
@@ -9410,11 +9410,11 @@ int llama_model_meta_val_str(const struct llama_model * model, const char * key,
...
     return (int)model->gguf_kv.size();
 }
 
-int llama_model_meta_key_by_index(const struct llama_model * model, int i, char * buf, size_t buf_size) {
+int32_t llama_model_meta_key_by_index(const struct llama_model * model, int i, char * buf, size_t buf_size) {
```
Comment (LoganDark) [inline llama.cpp:9417]: ssize_t

File: llama.cpp [hunk 9]
```diff
@@ -9426,7 +9426,7 @@ int llama_model_meta_key_by_index(const struct llama_model * model, int i, char
     return snprintf(buf, buf_size, "%s", it->first.c_str());
 }
 
-int llama_model_meta_val_str_by_index(const struct llama_model * model, int i, char * buf, size_t buf_size) {
+int32_t llama_model_meta_val_str_by_index(const struct llama_model * model, int32_t i, char * buf, size_t buf_size) {
```
Comment (LoganDark) [inline llama.cpp:9429]: ssize_t

File: llama.cpp [hunk 10]
```diff
@@ -9438,7 +9438,7 @@ int llama_model_meta_val_str_by_index(const struct llama_model * model, int i, c
     return snprintf(buf, buf_size, "%s", it->second.c_str());
 }
 
-int llama_model_desc(const struct llama_model * model, char * buf, size_t buf_size) {
+int32_t llama_model_desc(const struct llama_model * model, char * buf, size_t buf_size) {
```
Comment (LoganDark) [inline llama.cpp:9441]: ssize_t

File: llama.cpp [hunk 11]
```diff
@@ -9585,7 +9585,7 @@ void llama_kv_cache_view_update(const struct llama_context * ctx, struct llama_k
     }
 }
 
-int llama_get_kv_cache_token_count(const struct llama_context * ctx) {
+int32_t llama_get_kv_cache_token_count(const struct llama_context * ctx) {
```
Comment (LoganDark) [inline llama.cpp:9588]: size_t

File: llama.cpp [hunk 12]
```diff
@@ -9595,7 +9595,7 @@ int llama_get_kv_cache_token_count(const struct llama_context * ctx) {
     return result;
 }
 
-int llama_get_kv_cache_used_cells(const struct llama_context * ctx) {
+int32_t llama_get_kv_cache_used_cells(const struct llama_context * ctx) {
```
Comment (LoganDark) [inline llama.cpp:9598]: size_t

File: llama.cpp [hunk 13]
```diff
@@ -10075,7 +10075,7 @@ int llama_eval(
         struct llama_context * ctx,
                  llama_token * tokens,
                      int32_t   n_tokens,
-                         int   n_past) {
+                     int32_t   n_past) {
```
Comment (LoganDark) [inline llama.cpp:10078]: size_t (and n_tokens should also be size_t across the whole codebase)

File: llama.cpp [hunk 14]
```diff
@@ -10090,7 +10090,7 @@ int llama_eval_embd(
             struct llama_context * ctx,
                            float * embd,
                          int32_t   n_tokens,
-                             int   n_past) {
+                         int32_t   n_past) {
```
Comment (LoganDark) [inline llama.cpp:10093]: size_t

File: llama.cpp [hunk 15]
```diff
@@ -10233,12 +10233,12 @@ llama_token llama_token_eot(const struct llama_model * model) {
     return model->vocab.special_eot_id;
 }
 
-int llama_tokenize(
+int32_t llama_tokenize(
```
Comment (LoganDark) [inline llama.cpp:10236]: ssize_t

File: llama.cpp [hunk 16]
```diff
@@ -10266,7 +10266,7 @@ static std::string llama_decode_text(const std::string & text) {
 }
 
 // does not write null-terminator to buf
-int llama_token_to_piece(const struct llama_model * model, llama_token token, char * buf, int length) {
+int32_t llama_token_to_piece(const struct llama_model * model, llama_token token, char * buf, int32_t length) {
```
Comment (LoganDark) [inline llama.cpp:10269]: ssize_t

File: llama.h [hunk 17]
```diff
@@ -308,21 +308,20 @@ extern "C" {
...
-    // TODO: become more consistent with returned int types across the API
     LLAMA_API uint32_t llama_n_ctx      (const struct llama_context * ctx);
     LLAMA_API uint32_t llama_n_batch    (const struct llama_context * ctx);
 
     LLAMA_API enum llama_vocab_type llama_vocab_type(const struct llama_model * model);
 
-    LLAMA_API int llama_n_vocab    (const struct llama_model * model);
-    LLAMA_API int llama_n_ctx_train(const struct llama_model * model);
-    LLAMA_API int llama_n_embd     (const struct llama_model * model);
+    LLAMA_API int32_t llama_n_vocab    (const struct llama_model * model);
```
Comment (LoganDark) [inline llama.h:322]: size_t

File: llama.h [hunk 18]
```diff
@@ -308,21 +308,20 @@ extern "C" {
...
+    LLAMA_API int32_t llama_n_ctx_train(const struct llama_model * model);
```
Comment (LoganDark) [inline llama.h:323]: size_t

File: llama.h [hunk 19]
```diff
@@ -308,21 +308,20 @@ extern "C" {
...
+    LLAMA_API int32_t llama_n_embd     (const struct llama_model * model);
```
Comment (LoganDark) [inline llama.h:324]: size_t

File: llama.h [hunk 20]
```diff
@@ -333,19 +332,19 @@ extern "C" {
     // - GGUF array values are not supported by these functions
 
     // Get metadata value as a string by key name
-    LLAMA_API int llama_model_meta_val_str(const struct llama_model * model, const char * key, char * buf, size_t buf_size);
+    LLAMA_API int32_t llama_model_meta_val_str(const struct llama_model * model, const char * key, char * buf, size_t buf_size);
```
Comment (LoganDark) [inline llama.h:335]: ssize_t

File: llama.h [hunk 21]
```diff
@@ -333,19 +332,19 @@ extern "C" {
...
 
     // Get the number of metadata key/value pairs
-    LLAMA_API int llama_model_meta_count(const struct llama_model * model);
+    LLAMA_API int32_t llama_model_meta_count(const struct llama_model * model);
```
Comment (LoganDark) [inline llama.h:338]: size_t

File: llama.h [hunk 22]
```diff
@@ -333,19 +332,19 @@ extern "C" {
...
 
     // Get metadata key name by index
-    LLAMA_API int llama_model_meta_key_by_index(const struct llama_model * model, int i, char * buf, size_t buf_size);
+    LLAMA_API int32_t llama_model_meta_key_by_index(const struct llama_model * model, int32_t i, char * buf, size_t buf_size);
```
Comment (LoganDark) [inline llama.h:341]: ssize_t

File: llama.h [hunk 23]
```diff
@@ -333,19 +332,19 @@ extern "C" {
...
 
     // Get metadata value as a string by index
-    LLAMA_API int llama_model_meta_val_str_by_index(const struct llama_model * model, int i, char * buf, size_t buf_size);
+    LLAMA_API int32_t llama_model_meta_val_str_by_index(const struct llama_model * model, int32_t i, char * buf, size_t buf_size);
```
Comment (LoganDark) [inline llama.h:344]: ssize_t

File: llama.h [hunk 24]
```diff
@@ -333,19 +332,19 @@ extern "C" {
...
 
     // Get a string describing the model type
-    LLAMA_API int llama_model_desc(const struct llama_model * model, char * buf, size_t buf_size);
+    LLAMA_API int32_t llama_model_desc(const struct llama_model * model, char * buf, size_t buf_size);
```
Comment (LoganDark) [inline llama.h:347]: ssize_t

File: llama.h [hunk 25]
```diff
@@ -437,10 +436,10 @@ extern "C" {
 
     // Returns the number of tokens in the KV cache (slow, use only for debug)
     // If a KV cell has multiple sequences assigned to it, it will be counted multiple times
-    LLAMA_API int llama_get_kv_cache_token_count(const struct llama_context * ctx);
+    LLAMA_API int32_t llama_get_kv_cache_token_count(const struct llama_context * ctx);
```
Comment (LoganDark) [inline llama.h:439]: size_t

File: llama.h [hunk 26]
```diff
@@ -437,10 +436,10 @@ extern "C" {
...
 
     // Returns the number of used KV cells (i.e. have at least one sequence assigned to them)
-    LLAMA_API int llama_get_kv_cache_used_cells(const struct llama_context * ctx);
+    LLAMA_API int32_t llama_get_kv_cache_used_cells(const struct llama_context * ctx);
```
Comment (LoganDark) [inline llama.h:442]: size_t

File: llama.h [hunk 27]
```diff
@@ -531,7 +530,7 @@ extern "C" {
             struct llama_context * ctx,
                      llama_token * tokens,
                          int32_t   n_tokens,
-                             int   n_past),
+                         int32_t   n_past),
```
Comment (LoganDark) [inline llama.h:533]: size_t

File: llama.h [hunk 28]
```diff
@@ -540,7 +539,7 @@ extern "C" {
             struct llama_context * ctx,
                            float * embd,
                          int32_t   n_tokens,
-                             int   n_past),
+                         int32_t   n_past),
```
Comment (LoganDark) [inline llama.h:542]: size_t

File: llama.h [hunk 29]
```diff
@@ -633,24 +632,24 @@ extern "C" {
     /// @return Returns a negative number on failure - the number of tokens that would have been returned
     /// @param special Allow tokenizing special and/or control tokens which otherwise are not exposed and treated as plaintext.
     ///                Does not insert a leading space.
-    LLAMA_API int llama_tokenize(
+    LLAMA_API int32_t llama_tokenize(
```
Comment (LoganDark) [inline llama.h:635]: ssize_t

File: llama.h [hunk 30]
```diff
@@ -633,24 +632,24 @@ extern "C" {
...
         const struct llama_model * model,
                       const char * text,
-                             int   text_len,
+                         int32_t   text_len,
```
Comment (LoganDark) [inline llama.h:638]: size_t

File: llama.h [hunk 31]
```diff
@@ -633,24 +632,24 @@ extern "C" {
...
                      llama_token * tokens,
-                             int   n_max_tokens,
+                         int32_t   n_max_tokens,
```
Comment (LoganDark) [inline llama.h:640]: size_t

File: llama.h [hunk 32]
```diff
@@ -633,24 +632,24 @@ extern "C" {
...
                             bool   add_bos,
                             bool   special);
 
     // Token Id -> Piece.
     // Uses the vocabulary in the provided context.
     // Does not write null terminator to the buffer.
     // User code is responsible to remove the leading whitespace of the first non-BOS token when decoding multiple tokens.
-    LLAMA_API int llama_token_to_piece(
+    LLAMA_API int32_t llama_token_to_piece(
```
Comment (LoganDark) [inline llama.h:648]: ssize_t

File: llama.h [hunk 33]
```diff
@@ -633,24 +632,24 @@ extern "C" {
...
               const struct llama_model * model,
                            llama_token   token,
                                   char * buf,
-                                  int    length);
+                               int32_t   length);
```
Comment (LoganDark) [inline llama.h:652]: size_t

File: llama.h [hunk 34]
```diff
@@ -702,7 +701,7 @@ extern "C" {
     LLAMA_API void llama_sample_top_k(
             struct llama_context * ctx,
           llama_token_data_array * candidates,
-                             int   k,
+                         int32_t   k,
```
Comment (LoganDark) [inline llama.h:704]: size_t

File: llama.h [hunk 35]
```diff
@@ -761,7 +760,7 @@ extern "C" {
           llama_token_data_array * candidates,
                            float   tau,
                            float   eta,
-                             int   m,
+                         int32_t   m,
```
Comment (LoganDark) [inline llama.h:763]: size_t

File: llama.h [hunk 36]
```diff
@@ -834,8 +833,8 @@ extern "C" {
         llama_beam_search_callback_fn_t   callback,
                                    void * callback_data,
                                  size_t   n_beams,
-                                    int   n_past,
-                                    int   n_predict);
+                                int32_t   n_past,
```
Comment (LoganDark) [inline llama.h:836]: size_t

File: llama.h [hunk 37]
```diff
@@ -834,8 +833,8 @@ extern "C" {
...
+                                int32_t   n_predict);
```
Comment (LoganDark) [inline llama.h:837]: size_t

**Review by LoganDark — CHANGES_REQUESTED** [review]
Almost all of these need to be size_t or ssize_t instead since they represent the sizes of buffers in memory. Basically,

- buffer lengths need to be size_t.
- all indices need to be size_t.
- all counts need to be size_t (n_tokens, n_threads, n_gpu_layers etc)
- tokens themselves should be uint32_t rather than int32_t. going for size_t would double the memory cost of all tokens for no benefit since models don't yet have anywhere near 4 billion tokens.
- functions that return status codes can remain int32_t, as long as they don't smuggle a size in there (then they need to be ssize_t)
- the token encoding/decoding functions that can return a negative size in the case that the buffer is too small, need ssize_t.

This PR definitely doesn't address the majority of the issue in the codebase where int32_t has been chosen for values that really should be size_t or ssize_t instead. int32_t still has a place in some areas, but we surpassed 4GiB of memory decades ago.

disclaimer: I don't represent the llama.cpp project

**Comment by MarcusDunn** [comment]
I think changing to `int`s to `size_t`s as suggested would break things in `examples` and should be done with some care.

This PR is an attempt at some easy wins around making 32 bit `int` assumptions explicit - certainly not sufficient to call #4574 complete.

**Comment by LoganDark** [comment]
> I think changing to `int`s to `size_t`s as suggested would break things in `examples`

Is ABI compatibility a concern? Because if not we can just update their source code.

**Comment by MarcusDunn** [comment]
I was more thinking underflow.

**Comment by slaren** [comment]
There may be some argument between using signed or unsigned types, but I don't see any reason to use `size_t` for any of these. Sized integers are preferred to simplify interoperability. My opinion is that it is better to stick to signed integers unless we really need that last bit, since it often simplifies code logic and may result in faster code due to less strict overflow rules in C/C++.

**Comment by LoganDark** [comment]
> I was more thinking underflow.

Where exactly would there be underflow in any of the examples? You barely ever need signed computations for any of llama.cpp, the only thing I can think of is the tokenizer functions, which would be ssize_t which is fine.

**Comment by cebtenzzre** [comment]
> Almost all of these need to be size_t or ssize_t

[YAGNI](https://en.wikipedia.org/wiki/You_aren't_gonna_need_it).

**Comment by LoganDark** [comment]
> > Almost all of these need to be size_t or ssize_t
> 
> [YAGNI](https://en.wikipedia.org/wiki/You_aren't_gonna_need_it).

I know that practically, most of these are not going to need the integer precision, but *semantically* they still matter in certain contexts, such as Rust. I have three options:

- Force consumers of my bindings to do the casts to u32/i32 themselves.
- Silently truncate integers and introduce memory unsafety.
- Panic and abort the process if an integer would have been truncated.

None of these are particularly attractive, and while the third one may be most desirable from the downstream dependents' perspective, it's an absolute nightmare to pull off as a library author, because every conversion needs to be manually checked and panicked about.

Just use the proper integer types please.

**Review by ggerganov — APPROVED** [review]

**Pull request merged**